master/cache.lua

```diff
@@ -0,0 +1,465 @@
+local cjson_safe = require "cjson.safe"
+local shmlock = require "resty.lock"
+local http = require "resty.http.simple"
+
+
+local _M = {}
+
+-- In order to make caching code testable, these constants need to be
+-- configurable/exposed through env vars.
+--
+-- Values assigned to these variable need to fufil following condidtion:
+--
+-- CACHE_FIRST_POLL_DELAY << CACHE_EXPIRATION < CACHE_POLL_PERIOD < CACHE_MAX_AGE_SOFT_LIMIT < CACHE_MAX_AGE_HARD_LIMIT
+--
+-- CACHE_BACKEND_REQUEST_TIMEOUT << CACHE_REFRESH_LOCK_TIMEOUT
+--
+-- All are in units of seconds. Below are the defaults:
+local _TIMERS = {}
+local env_vars = {CACHE_FIRST_POLL_DELAY = 2,
+                  CACHE_POLL_PERIOD = 25,
+                  CACHE_EXPIRATION = 20,
+                  CACHE_MAX_AGE_SOFT_LIMIT = 35,
+                  CACHE_MAX_AGE_HARD_LIMIT = 259200,
+                  CACHE_BACKEND_REQUEST_TIMEOUT = 10,
+                  CACHE_REFRESH_LOCK_TIMEOUT = 20,
+                  }
+
+for key, value in pairs(env_vars) do
+    -- yep, we are OK with key==nil, tonumber will just return nil
+    local env_var_val = tonumber(os.getenv(key))
+    if env_var_val == nil or env_var_val == value then
+        ngx.log(ngx.DEBUG, "Using default ".. key .. " value: `" .. value .. "` seconds")
+        _TIMERS[key] = value
+    else
+        ngx.log(ngx.NOTICE,
+                key .. " overridden by ENV to `" .. env_var_val .. "` seconds")
+        _TIMERS[key] = env_var_val
+    end
+end
+
+
+local function cache_data(key, value)
+    -- Store key/value pair to SHM cache (shared across workers).
+    -- Return true upon success, false otherwise.
+    -- Expected to run within lock context.
+
+    local cache = ngx.shared.cache
+    local success, err, forcible = cache:set(key, value)
+    if success then
+        return true
+    end
+    ngx.log(
+        ngx.ERR,
+        "Could not store " .. key .. " to state cache: " .. err
+        )
+    return false
+end
+
+
+local function request(host, port, path, accept_404_reply)
+    -- Use cosocket-based HTTP library, as ngx subrequests are not available
+    -- from within this code path (decoupled from nginx' request processing).
+    -- The timeout parameter is given in milliseconds.
+    local res, err = http.request(host, port,
+        {
+            path = path,
+            timeout = _TIMERS.CACHE_BACKEND_REQUEST_TIMEOUT * 1000,
+        }
+    )
+
+    if not res then
+        return nil, err
+    end
+
+    if res.status ~= 200 and (accept_404_reply and res.status ~= 404 or not accept_404_reply) then
```

Comment:
I added one more nesting level. `accept_404_reply and res.status ~= 404 or not accept_404_reply` will have to stay as-is unless we want to duplicate code.